### ISSUE TITLE
Cannot remove old versions

### ISSUE THREAD

**Issue by henryx** [johanjanssen/ChocolateyPackages#5]
Upgrading from OpenJDK 11 to OpenJDK 12, old version not only is not uninstalled but is not showed with `choco list -lai`

**Comment by johanjanssen** [comment]
I tried to reproduce your remark with the steps below. But it seems to work correct on my machine. Do I miss something? Or maybe I don't understood your issue correctly.

choco install openjdk --version 11.0.2.01
choco upgrade openjdk

After this command 11.0.2.01 is removed from C:\Program Files\OpenJDK and choco list -lai gives:
openjdk 12.0.0
.. packages installed.

**Comment by henryx** [comment]
I've followed these steps:
 - `choco install openjdk --version 11.0.2.01` works correctly
 - `choco list -lai` shows `openjdk 11.0.2.01`
 - `choco upgrade openjdk`
 - `choco list -lai` show *only* `openjdk 12.0.0`
 - `choco install openjdk --version 11.0.2.01`  return error

So, only mode for removing OpenJDK 11 is by hand

**Comment by johanjanssen** [comment]
Sorry apparently I wasn't completely awake when I tried the first time.

Indeed when I follow your steps and look at the OpenJDK folder on disk I see two versions: 11.0.2.01 and 12.0.0. However until now I saw that as a feature, so people have the choice to keep the old version next to the new version. But I completely understand that you could also see it as a bug.

To remove the old version you could do a `choco uninstall openjdk` just before the `choco upgrade openjdk`. I understand that it's an extra step, but this way I can easily support both options:

1. Keeping multiple versions of OpenJDK
2. Keeping just one version of OpenJDK

What do you think of that?

**Comment by henryx** [comment]
A good choice is to maintain multiple versions of OpenJDK (I understand that the Java world is more complicated on this aspect). Chocolatey can do this natively, and is the preferable way. So, the good way can be as the `choco list -lai` command shows all versions of the OpenJDK installed. In this way, if you want to remove old versions, you can use `choco uninstall openjdk --version 11.0.2.01`

**Comment by johanjanssen** [comment]
That sounds like a good idea. Do you maybe know how I can achieve that in my configuration?

**Comment by henryx** [comment]
Sadly,  no :(

**Comment by metametadata** [comment]
I experience the similar problem with GraalVM package.

## Steps

1) Install the specific GraalVM version, e.g. `choco install -y graalvm --version 20.2.0`. This works:

```
> java -version
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment GraalVM CE 20.2.0 (build 11.0.8+10-jvmci-20.2-b03)
OpenJDK 64-Bit Server VM GraalVM CE 20.2.0 (build 11.0.8+10-jvmci-20.2-b03, mixed mode, sharing)

> (Get-Command -CommandType Application -Name java").Definition
C:\Program Files\GraalVM\graalvm-ce-java11-20.2.0\bin\java.exe

> $env:path
C:\Python38\Scripts\;C:\Python38\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\ProgramData\chocolatey\bin;C:\Program Files\Git\cmd;C:\Program Files\GraalVM\graalvm-ce-java11-20.2.0\bin;
```

2) Install another version, e.g. `choco install -y graalvm --version 20.3.0`.

# Expected

`java` points to 20.3.0, only a single JDK is installed, PATH contains a single GraalVM path.

# Actual

`java` points to 20.2.0, two JDKs are in `Program Files`, `PATH` contains both paths:

```
> java -version
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment GraalVM CE 20.2.0 (build 11.0.8+10-jvmci-20.2-b03)
OpenJDK 64-Bit Server VM GraalVM CE 20.2.0 (build 11.0.8+10-jvmci-20.2-b03, mixed mode, sharing)

> $env:path                                                                                         
C:\Python38\Scripts\;C:\Python38\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\ProgramData\chocolatey\bin;C:\Program Files\Git\cmd;C:\Program Files\GraalVM\graalvm-ce-java11-20.2.0\bin;C:\Program Files\GraalVM\graalvm-ce-java11-20.3.0\bin;

> (Get-Command -CommandType Application -Name"java").Definition
C:\Program Files\GraalVM\graalvm-ce-java11-20.2.0\bin\java.exe
C:\Program Files\GraalVM\graalvm-ce-java11-20.3.0\bin\java.exe
```

**Comment by johanjanssen-sanoma** [comment]
@metametadata 
Until now I saw that as a feature, so people have the choice to keep the old version next to the new version. But I completely understand that you could also see it as a bug.

To remove the old version you could do a choco uninstall openjdk just before the choco upgrade openjdk. I understand that it's an extra step, but this way I can easily support both options:

- Keeping multiple versions of OpenJDK
- Keeping just one version of OpenJDK

Only thing I don't understand is that Java points to the old version. Did you try the command in a new shell?

**Comment by metametadata** [comment]
> To remove the old version you could do a choco uninstall openjdk just before the choco upgrade openjdk.

Yes, this helps. But I suppose this is unexpected because the idiomatic way is to use `-m` flag?

> -m, --sxs, --sidebyside, --side-by-side, --allowmultiple, --allow-multiple, --allowmultipleversions, --allow-multiple-versions
>     AllowMultipleVersions - Should multiple versions of a package be
>       installed? Defaults to false.

> Only thing I don't understand is that Java points to the old version. Did you try the command in a new shell?

Correct. It seems to be connected to the PATH because reordering the paths by hand in GUI makes it find the newer `java`:

```
> java -version
openjdk version "11.0.9" 2020-10-20                                                   
OpenJDK Runtime Environment GraalVM CE 20.3.0 (build 11.0.9+10-jvmci-20.3-b06)
OpenJDK 64-Bit Server VM GraalVM CE 20.3.0 (build 11.0.9+10-jvmci-20.3-b06, mixed mode, sharing) 

> $env:path                                                                                                                                                                              
C:\Python38\Scripts\;C:\Python38\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\ProgramData\chocolatey\bin;C:\Program Files\Git\cmd;C:\Program Files\GraalVM\graalvm-ce-java11-20.3.0\bin;C:\Program Files\GraalVM\graalvm-ce-java11-20.2.0\bin;

> (Get-Command -CommandType Application -Name "java").Definition
C:\Program Files\GraalVM\graalvm-ce-java11-20.3.0\bin\java.exe
C:\Program Files\GraalVM\graalvm-ce-java11-20.2.0\bin\java.exe
```

**Comment by johanjanssen-sanoma** [comment]
> > To remove the old version you could do a choco uninstall openjdk just before the choco upgrade openjdk.
> 
> Yes, this helps. But I suppose this is unexpected because the idiomatic way is to use `-m` flag?

That could be, however support for that flag isn't working out of the box I guess. I assume in my scripts I need to handle that flag and build the appropriate logic?

If someone wants to do that, feel free to create a pull request. If the script isn't too complex/complicated to maintain, then I'm more then willing to merge it.

> > Only thing I don't understand is that Java points to the old version. Did you try the command in a new shell?
> 
> Correct. It seems to be connected to the PATH because reordering the paths by hand in GUI makes it find the newer `java`:

Weird normally I've seen that the newest installation path is placed at the beginning of the PATH and thereby loaded first if you execute Java. If I've got some time, then I'll try and test it to see if the behavior is the same on my machine.

**Comment by metametadata** [comment]
> I assume in my scripts I need to handle that flag and build the appropriate logic?

Hard to tell. I've only used Chocolatey as a user.

> If I've got some time, then I'll try and test it to see if the behavior is the same on my machine.

OK, thanks. Just in case, I experimented in Windows 2012. Versions of Chocolatey and Powershell:

```
> $PSVersionTable.PSVersion                                                                                                                                                    
                                                                                                                                                                                                            
Major  Minor  Build  Revision                                                                                                                                                                               
-----  -----  -----  --------                                                                                                                                                                               
4      0      -1     -1

> choco version                                                                                                                                                                          
Chocolatey v0.10.15
...
```
